File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0 for _ in range(n + 2)]

        for i in range(n - 1, -1, -1):
            take = nums[i] + dp[i + 2]
            dont_take = dp[i + 1]
            dp[i] = max(take, dont_take)
        
        return dp[0]
        
        @cache
        def dp(i: int):
            if i >= len(nums):
                return 0
            take = nums[i] + dp(i + 2)
            not_take = dp(i + 1)
            return max(take, not_take)
        
        return dp(0)

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        int n = nums.size();
        function<int(int)> find_house_robbed = [&] (int i) -> int {
            if (i >= n) return 0;
            if (dp[i] != -1) return dp[i];

            dp[i] = max(find_house_robbed(i + 1), nums[i] + find_house_robbed(i + 2));
            return dp[i];
        };

        return find_house_robbed(0);
    }
};